OVERVIEW.md

```diff
@@ -0,0 +1,102 @@
+
+This service is usable in a variety of environments but it has been configured to run in a Rancher system. The rancher configuration is a `docker-compose.yml` file at its core, with some configuration options for Rancher. The service starts with these components:
+
+- `demoservice-lb`: Load balancer to route traffic to web services
+- App runs as two services
+  - `demoservice-web`: Django frontend service
+  - `demoservice-worker`: Celery worker
+- `demoservice-db`: Postgres database
+- `demoservice-rabbit`: RabbitMQ database
+
+Separately to this, a [Traefik](https://traefik.io/) instance runs to route traffic. This will be explained a little more clearly later in the document.
+
+
+### App structure
+
+A simple overview of the folder structure:
+
+```
+├── app
+│   └── demoservice
+│       ├── libs
+│       │   └── [DJANGO/PYTHON HELPER FILES]
+│       ├── migrations
+│       │   └── [DJANGO DATABASE MIGRATION SCRIPTS]
+│       └── templates
+│           └── [DJANGO HTML TEMPLATES]
+├── bin
+│   └── [HELPER SCRIPTS]
+└── templates
+    └── [RANCHER DEPLOYMENT TEMPLATE]
+```
+
+Inside the demoservice folder there are a couple of base files:
+- `tasks.py`
+  - Celery tasks and configuration
+- `views.py`
+  - webhook endpoint
+  - frontend management pages
+
+
+## Flow
+
+### Webhooks
+- Webhook, important to check secret
+ - Listens for pull_request events of type 'opened', 'closed', and 'synchronised'.
+- ...
+- Starts ./run demo
+
+_**When receiving webhooks, it is important to always check the secret hash. ALWAYS.**_
+
+#### Github webhooks
+Github webhooks are pushed to the `/webhook/github` path. When the app receives a payload from GitHub it will check the signature against a hash made from our secret. We must check this to ensure that only webhooks from verified sources are started.
+
+Github can send a variety of event types. The type is checked against the header `HTTP_X_GITHUB_EVENT`. This then checks the type of event in the payload content. We primarily listen for `opened`, `closed`, and `synchronized`. Synchronized is triggered when a branch has been updated; often by rebasing or pushing new commits.
+
+When the webhook has been understood, it queues up a relevant task in the message queue. It will either start/update a demo or delete one. Starting a demo will stop any previously running demos, so this action can be used for updating a demo.
+
+### Frontend management
+The demo service has a simple admin/management interface for viewing demos, as well as starting and stopping them. This interface requires an openid login and this should be enforced to prevent unauthorised users creating demos on the system.
+
+Running demos are displayed by filtering the Docker containers on the machine with the relevant `run.demo=true` label.
```

Comment:
Not sure what do you mean by that, could you add links to show where this comes from?

Reply:
Yes I think a command-line snippet would help illustrate this point.

Reply:
Good points, updated